File: destack-py/destack/test/language/struct.py
from destack.language import (
    CheckedType,
    PropertyDefinition,
    Session,
    Struct,
    StructType,
    Type,
)


def test_struct_inheritance(session: Session):
    """Test the Struct inheritance hierarchy."""
    # test Struct as base
    assert Struct.metatype == StructType.STRUCT
    assert Struct.__definition__.is_abstract

    # test Type extends Struct
    assert Type.metatype == StructType.TYPE
    assert not Type.__definition__.is_abstract
    assert Type.__definition__.base_type == StructType.STRUCT
    assert Type.__definition__.inherits == [StructType.STRUCT]

    # test CheckedType extends Type
    assert CheckedType.metatype == StructType.CHECKED_TYPE
    assert not CheckedType.__definition__.is_abstract
    assert CheckedType.__definition__.base_type == StructType.TYPE
    assert CheckedType.__definition__.inherits == [StructType.STRUCT, StructType.TYPE]

    # test PropertyDefinition extends CheckedType
    assert PropertyDefinition.metatype == StructType.PROPERTY_DEFINITION
    assert not PropertyDefinition.__definition__.is_abstract
    assert PropertyDefinition.__definition__.base_type == StructType.CHECKED_TYPE
    assert PropertyDefinition.__definition__.inherits == [
        StructType.STRUCT,
        StructType.TYPE,
        StructType.CHECKED_TYPE,
    ]

File: destack-rs/destack/src/core/definition/struct.rs
//! destack.core.definition.struct@2025.08.15.1

#![destack::partial(destack.core.definition.struct, file)]

use crate::{
    ConstantDefinition, MethodDefinition, NodeType, ObjectStability, PropertyDefinition,
    StructType, TagDefinition, UniverseCategory, UniverseDomain,
};

#[destack::generated(StructDefinition, -, block)]
/// Definition of a builtin Struct.
pub struct StructDefinition {
    pub id: u32,
    pub r#type: StructType,
    pub name: String,
    pub description: String,
    pub domain: UniverseDomain,
    pub category: UniverseCategory,
    pub stability: ObjectStability,
    pub taggings: Vec<u8>,
    pub is_immutable: bool,
    pub is_abstract: bool,
    pub is_interned: bool,
    pub properties: Vec<PropertyDefinition>,
    pub methods: Vec<MethodDefinition>,
    pub constants: Vec<ConstantDefinition>,
    pub tags: Vec<TagDefinition>,
    pub base_type: Option<StructType>,
    pub extended_by: Vec<StructType>,
    pub inherits: Vec<StructType>,
    pub inherited_by: Vec<StructType>,
    pub into_node_types: Vec<NodeType>,
}
